src/routes/analysis.py

```
@@ -3,273 +3,390 @@
+            concept = Concept(
+                document_id=doc.id,
+                name=c['name'],
+                category=c.get('category'),
+                importance=c.get('importance'),
+                difficulty=c.get('difficulty'),
+                description=c.get('description')
+            )
+            db.session.add(concept)
+            db.session.flush()
+            # attach exercises
+            for ex in [e for e in exercises if e['concept'] == c['name']][0]['exercises']:
```

Ruff (0.13.1)

Prefer `next(e for e in exercises if e['concept'] == c['name'])` over single element slice

Replace with `next(e for e in exercises if e['concept'] == c['name'])`

(RUF015)